[compiletest] Ignore known paths when abbreviating output

pietroalbini opened the pull request:
To prevent out of memory conditions, compiletest limits the amount of output a test can generate, abbreviating it if the test emits more than a threshold. While the behavior is desirable, it also causes some issues (like #96229, #94322 and #92211).

The latest one happened recently, when the `src/test/ui/numeric/numeric-cast.rs` test started to fail on systems where the path of the rust-lang/rust checkout is too long. This includes my own development machine and [LLVM's CI](https://github.com/rust-lang/rust/issues/96362#issuecomment-1108609893). Rust's CI uses a pretty short directory name for the checkout, which hides these sort of problems until someone runs the test suite on their own computer.

When developing the fix I tried to find the most targeted fix that would prevent this class of failures from happening in the future, deferring the decision on if/how to redesign abbreviation to a later date. The solution I came up with was to ignore known base paths when calculating whether the output exceeds the abbreviation threshold, which removes this kind of nondeterminism.

This PR is best reviewed commit-by-commit.

rust-highfive commented:
r? @Mark-Simulacrum

(rust-highfive has picked a reviewer for you, use r? to override)

Mark-Simulacrum commented on src/tools/compiletest/src/read2.rs:
```diff
@@ -3,31 +3,54 @@
+                    // To make those failures deterministic across all environments we ignore known
+                    // paths when calculating the string length, while still including the full
+                    // path in the output. This could result in some output being larger than the
+                    // threshold, but it's better than having nondeterministic failures.
```
#This also defeats the OOM protection, if we have for some reason rustc being over eager with *just* outputting paths. (Or repeating a path). It also won't avoid non-determinism in terms of failures here, at least not those caused by other differences in behavior (e.g., cfg(unix) vs not differences if testing on Windows).

Right now we abbreviate output after 416 KB of output from a single test. It seems like OOM-wise we should be comfortable bumping to, say, 5 megabytes -- I would expect any "reasonable" output from rustc to fit into that for our tests. That should make path length differences less of an issue, IMO, and is a much simpler change and feels less scary.

If we want to keep an eye on the tests so we catch cases where we dump kilobytes into the stderr for no good reason, I think our best bet would be a system closer to perf.rust-lang.org or toolstate; I could pretty easily imagine us dumping each test's output size (and maybe runtime?) into a file and then having a simple comparison post-merge that we're not increasing more than 10% (or staying in some <512KB range or something). I'm not really convinced we actually need to do anything here though.

the8472 replied:
> This also defeats the OOM protection, if we have for some reason rustc being over eager with just outputting paths. (Or repeating a path).

It would have to output just the source or build or source path repeatedly, not a subpath thereof. Otherwise the subpaths would still be counted and eventually exceed the size limit.

pietroalbini replied:
> This also defeats the OOM protection, if we have for some reason rustc being over eager with just outputting paths. (Or repeating a path).

Pushed a tentative second commit addressing this, by adding a known placeholder of 32 bytes for each excluded pattern. Changing the placeholder to 64 triggers the abbreviation btw, so it seems like a good enough value to keep the behavior close to the current one.

> It also won't avoid non-determinism in terms of failures here, at least not those caused by other differences in behavior (e.g., cfg(unix) vs not differences if testing on Windows).

Sure! It would still prevent some non-determinism, and other kinds like the ones you mention would likely be catched by the CI builders we have.

> Right now we abbreviate output after 416 KB of output from a single test. It seems like OOM-wise we should be comfortable bumping to, say, 5 megabytes -- I would expect any "reasonable" output from rustc to fit into that for our tests. That should make path length differences less of an issue, IMO, and is a much simpler change and feels less scary.

Hmm, while 5MB would likely avoid the issue I'm facing right now, it also wouldn't have catched #94327, which seems like a category of issues worth fixing. Also, the (fairly short) rustfix test in question (`src/test/ui/numeric/numeric-cast.rs`) emits 793 paths in its JSON output, and I could see more complex rustfix tests in the future, emitting way more diagnostics (and thus more paths) than that.

> If we want to keep an eye on the tests so we catch cases where we dump kilobytes into the stderr for no good reason, I think our best bet would be a system closer to perf.rust-lang.org or toolstate; I could pretty easily imagine us dumping each test's output size (and maybe runtime?) into a file and then having a simple comparison post-merge that we're not increasing more than 10% (or staying in some <512KB range or something). I'm not really convinced we actually need to do anything here though.

This seems way more complex than the solution proposed PR, and would also be harder to integrate for someone running CIs outside of the Rust project.

the8472 replied:
I don't think it's necesasry, but now that the safety net is there it doesn't hurt either as long as CI and local builds are equal before this judge.

pietroalbini commented on src/tools/compiletest/src/read2.rs:
```diff
@@ -12,28 +12,32 @@ pub struct ReadOutput {
-                    bytes.extend_from_slice(data);
-                    let new_len = bytes.len();
+                ProcOutput::Full(ref mut string) => {
+                    let utf8 = String::from_utf8_lossy(data);
```
Good catch, pushed a new (simpler) implementation that works over bytes instead of strings, and removed the refactors required to handle strings inside `read2`.

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6231023193?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/6ad414f73b2ba8a6796646adc8ed11c6669d16ce/checks/6231023193/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
Diff in /checkout/src/tools/compiletest/src/runtest.rs at line 1791:
             child.stdin.as_mut().unwrap().write_all(input.as_bytes()).unwrap();
 
 
-        let Output { status, stdout, stderr } =
-            self.read2_abbreviated(child);
+        let Output { status, stdout, stderr } = self.read2_abbreviated(child);
         let result = ProcRes {
             status,
             status,
Running `"/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/rustfmt" "--config-path" "/checkout" "--edition" "2021" "--unstable-features" "--skip-children" "--check" "/checkout/src/tools/compiletest/src/errors.rs" "/checkout/src/tools/compiletest/src/runtest.rs" "/checkout/src/tools/compiletest/src/header/tests.rs" "/checkout/src/tools/compiletest/src/util.rs" "/checkout/src/tools/compiletest/src/compute_diff.rs" "/checkout/src/tools/html-checker/main.rs" "/checkout/src/tools/compiletest/src/raise_fd_limit.rs" "/checkout/src/tools/compiletest/src/util/tests.rs"` failed.
If you're running `tidy`, try again with `--bless`. Or, if you just want to format code, run `./x.py fmt` instead.
```

</details>

Mark-Simulacrum commented on src/tools/compiletest/src/read2.rs:
```diff
@@ -5,29 +5,53 @@ pub use self::imp::read2;
+                            .filter(|window| window == &pattern_bytes)
+                            .count();
+                        *excluded_len += matches as isize
+                            * (EXCLUDED_PLACEHOLDER_LEN - pattern_bytes.len() as isize);
```
This approach won't really work, I think, at least not in general. The data we receive here is an ~arbitrary slice of the total test output, which means that there's no guarantee that the pattern we're looking for isn't lying across a boundary (i.e., half the pattern in one "chunk" and half in the next).

I think the way to mitigate this would be to start matching here pattern_bytes-1 into the already pushed data, but I'm not entirely confident on the full correctness of that approach. (It may also be worth considering a RegexSet here, from a performance perspective, to do a single pass over the incoming bytes. But with our hard upper bound being pretty small it's probably not worth it).

Mark-Simulacrum commented:
I'm OK with approving this with the nit fixed. Still feels a little unfortunate/off to me, but I'm not seeing massively better options, particularly if there's a soft goal of trying to prevent 'excessively large' outputs from rustc. (Not that several hundred kilobytes of JSON isn't already quite big for the ~1 KB input we typically have...).

pietroalbini commented:
Addressed the last suggestion you wrote. When fixing it, the amount of length manipulations we're doing started worrying me without tests, so I added them in another commit.

Those tests also uncovered an existing bug (if the first `extend` was longer than the maximum size it would not be truncated, but its excess size would still be counted for the "skipped n bytes" message), which I then fixed :slightly_smiling_face:

Mark-Simulacrum commented on src/tools/compiletest/src/read2.rs:
```diff
@@ -74,6 +27,87 @@ pub fn read2_abbreviated(mut child: Child) -> io::Result<Output> {
+                        .filter(|window| window == &pattern_bytes)
+                        .count();
+                    *excluded_len += matches as isize
+                        * (EXCLUDED_PLACEHOLDER_LEN - pattern_bytes.len() as isize);
```
I think the EXCLUDED_PLACEHOLDER_LEN is pretty confusing to me (why 32?) -- it seems like the effect here is that we're going to sometimes shrink and sometimes increase the excluded_len depending on the pattern_bytes? It's hard for me to conceptualize what the effect of that will be in the long run.

Maybe a better/simpler approach here is to set an upper bound on the excluded bytes (say, 2 MB, or so) which seems like it would more clearly express that while we do want to avoid different failures across different systems, we probably also want to avoid enormous outputs from rustc. I think the effect of the placeholder length here is probably pretty similar regardless (right?), just less obviously correct.

pietroalbini replied:
The effect of this is that all paths will be considered 32 bytes long, regardless of whether they're shorter or longer. This should remove all path size differences between different machines.

I chose 32 bytes because that's the highest power of two that passes all the tests. Specifically for `src/test/ui/numeric/numeric-cast.rs`, all path lengths greater than 53 cause that test to fail. If you want the behavior to be more consistent with today I guess we could go with 48?

Mark-Simulacrum replied:
Why do we need to subtract pattern_bytes from EXCLUDED_PLACEHOLDER_LEN though? It looks like we have a total match count (`matches`) and if we want to ensure that no matter what the pattern is we'd count it as 32 bytes, we'd just multiply by 32 (`EXCLUDED_PLACEHOLDER_LEN`) here, right?

32 is fine with this rationale, but I do think we should put a comment on its definition saying something like "Number of bytes any excluded segment of the input will be treated as", with perhaps a short note for why (cross-platform consistency).

pietroalbini replied:
> Why do we need to subtract pattern_bytes from EXCLUDED_PLACEHOLDER_LEN though? It looks like we have a total match count (matches) and if we want to ensure that no matter what the pattern is we'd count it as 32 bytes, we'd just multiply by 32 (EXCLUDED_PLACEHOLDER_LEN) here, right?

No, that'd still cause nondeterminism (`excluded_len` is the number of bytes that should be ignored from the output size).

The `pattern` is the path on the local filesystem (such as `/home/ci/rust`), so if we want to make things deterministic we need to ignore the full length of the pattern when counting the string size (so `matches * pattern_bytes.len()`).

That works, but it would lead to OOM if the compiler emits only the patterns without any extra bytes, as the whole output size would be ignored. To avoid that the code subtracts 32 from the pattern length: when the length of the pattern is lower than 32 the output will be considered *longer* (to simulate a system with longer paths), while when the length of the pattern is greater than 32 bytes it will be considered *shorter* (removing the nondeterminism, while preventing OOMs due to the 32 bytes still being counted).

In practice that's `32 - pattern_bytes` rather than `pattern_bytes - 32` due to the code below doing `len + excluded_bytes` rather than `len - excluded_bytes`.

I can change the code to invert signs (making things more clear) and add another comment explaining this.

Mark-Simulacrum replied:
Ah, I think part of my confusion was that pattern is a local file system path (rather than something more abstract like a regex, e.g., "/[^"]+)

I think we should maybe consider renaming it to something like path_prefix since it isn't really a pattern - or, at least, doesn't match more than one specific str.

Mark-Simulacrum replied:
I think the best thing here is to subtract `matches * pattern_bytes.len()` and then add `matches * 32`, which should I think be equivalent and much more obvious (to me, anyway), in terms of effect.

pietroalbini replied:
Pushed a commit addressing this.

Mark-Simulacrum commented on src/tools/compiletest/src/read2/tests.rs:
```diff
@@ -0,0 +1,123 @@
+    out.extend(&data, &[]);
+
+    let mut expected = vec![b'.'; HEAD_LEN];
+    expected.extend_from_slice(b"\n\n<<<<<< SKIPPED 16 BYTES >>>>>>\n\n");
```
FWIW, I think we should make the read2_abbreviated return an error or panic instead of inserting this skipped message -- that typically leads to a JSON de-serialization failure and a pretty opaque one at that, which I think isn't too helpful -- we can probably say something more useful like "captured X bytes, which is above the limit in compiletest. See abbreviated output: {output}".

pietroalbini replied:
Yeah, let's do that in a separate PR though.

pietroalbini commented:
@bors review

rust-log-analyzer commented:
The job **`mingw-check`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/runs/6739656252?check_suite_focus=true) [(plain)](https://github.com/rust-lang/rust/commit/e257f38160b361f738e9c42c3b76665fbbf13aa4/checks/6739656252/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
configure: rust.debug-assertions := True
configure: rust.overflow-checks := True
configure: llvm.assertions      := True
configure: dist.missing-tools   := True
configure: build.configure-args := ['--enable-sccache', '--disable-manage-submodu ...
configure: writing `config.toml` in current directory
configure: 
configure: run `python /checkout/x.py --help`
configure: 
---
   Compiling compiletest v0.0.0 (/checkout/src/tools/compiletest)
error[E0432]: unresolved import `crate::read2::EXCLUDED_PLACEHOLDER_LEN`
 --> src/tools/compiletest/src/read2/tests.rs:1:32
  |
1 | use crate::read2::{ProcOutput, EXCLUDED_PLACEHOLDER_LEN, HEAD_LEN, TAIL_LEN};
  |                                ^^^^^^^^^^^^^^^^^^^^^^^^ no `EXCLUDED_PLACEHOLDER_LEN` in `read2`

error[E0026]: variant `ProcOutput::Full` does not have a field named `excluded_len`
   |
   |
73 |         ProcOutput::Full { excluded_len, .. } => assert_eq!(
   |                            ^^^^^^^^^^^^ variant `ProcOutput::Full` does not have this field
Some errors have detailed explanations: E0026, E0432.
For more information about an error, try `rustc --explain E0026`.
error: could not compile `compiletest` due to 2 previous errors
Build completed unsuccessfully in 0:01:24
```

</details>

Mark-Simulacrum commented on src/tools/compiletest/src/read2.rs:
```diff
@@ -58,21 +65,25 @@ impl ProcOutput {
                         .count();
-                    *excluded_len += matches as isize
-                        * (EXCLUDED_PLACEHOLDER_LEN - pattern_bytes.len() as isize);
+                    *filtered_len -= matches * path_bytes.len();
```
I think this is going to run into some problems if we have overlapping matches (e.g., one of the patterns is /home/mark and the other is /home/mark/rust, then we'll subtract out twice).

I think src_base and build_base won't do that in practice so we're OK for now, but ideally we'd either (a) assert that it doesn't happen or (b) somehow fix it. I'm not quite sure what the best approach to fixing it would be -- trying to detect this kind of overlap seems annoying.

Mark-Simulacrum commented:
r=me if you don't want to deal with this additional problem, but maybe add a comment about it to where we define paths so we think about it.

pietroalbini commented:
@bors r=Mark-Simulacrum

[There is already a comment about that where the filtered paths are defined.](https://github.com/rust-lang/rust/pull/96551/files#diff-03a0567fa80ca04ed5a55f9ac5c711b4f84659be2d0ac4a984196d581c04f76bR1749-R1753)

bors commented:
:pushpin: Commit 8ea95988df4d493ec3f556ab0de6b0b812ed27be has been approved by `Mark-Simulacrum`

<!-- @bors r=Mark-Simulacrum 8ea95988df4d493ec3f556ab0de6b0b812ed27be -->
<!-- homu: {"type":"Approved","sha":"8ea95988df4d493ec3f556ab0de6b0b812ed27be","approver":"Mark-Simulacrum"} -->

bors commented:
:hourglass: Testing commit 8ea95988df4d493ec3f556ab0de6b0b812ed27be with merge 6609c6734de4df43e24d7672f8ae8786ecc8047e...
<!-- homu: {"type":"BuildStarted","head_sha":"8ea95988df4d493ec3f556ab0de6b0b812ed27be","merge_sha":"6609c6734de4df43e24d7672f8ae8786ecc8047e"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/runs/6752066376?check_suite_focus=true)
Approved by: Mark-Simulacrum
Pushing 6609c6734de4df43e24d7672f8ae8786ecc8047e to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Mark-Simulacrum","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/runs/6752066376?check_suite_focus=true"},"merge_sha":"6609c6734de4df43e24d7672f8ae8786ecc8047e"} -->

rust-timer commented:
Finished benchmarking commit (6609c6734de4df43e24d7672f8ae8786ecc8047e): [comparison url](https://perf.rust-lang.org/compare.html?start=760237ff785fd14ac7fdab799f4d695d86cf9cbf&end=6609c6734de4df43e24d7672f8ae8786ecc8047e&stat=instructions:u).


### [Instruction count](https://perf.rust-lang.org/compare.html?start=760237ff785fd14ac7fdab799f4d695d86cf9cbf&end=6609c6734de4df43e24d7672f8ae8786ecc8047e&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=760237ff785fd14ac7fdab799f4d695d86cf9cbf&end=6609c6734de4df43e24d7672f8ae8786ecc8047e&stat=max-rss)
<details>
<summary>Results</summary>


- Primary benchmarks: mixed results
- Secondary benchmarks: 😿 relevant regression found

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 0.1%     | 0.1% | 1         |
| Regressions 😿 <br /> (secondary) | 2.4%     | 2.4% | 1         |
| Improvements 🎉 <br /> (primary) | -3.0%    | -3.0% | 1         |
| Improvements 🎉 <br /> (secondary) | N/A      | N/A | 0         |
| All 😿🎉 (primary) | -1.4%    | -3.0% | 2         |
</details>

### [Cycles](https://perf.rust-lang.org/compare.html?start=760237ff785fd14ac7fdab799f4d695d86cf9cbf&end=6609c6734de4df43e24d7672f8ae8786ecc8047e&stat=cycles:u)
<details>
<summary>Results</summary>


- Primary benchmarks: 😿 relevant regression found
- Secondary benchmarks: mixed results

|            | mean[^1] | max | count[^2] |
|:----------:|:--------:|:---:|:---------:|
| Regressions 😿 <br /> (primary) | 3.0%     | 3.0% | 1         |
| Regressions 😿 <br /> (secondary) | 2.2%     | 2.2% | 1         |
| Improvements 🎉 <br /> (primary) | N/A      | N/A | 0         |
| Improvements 🎉 <br /> (secondary) | -3.6%    | -3.6% | 1         |
| All 😿🎉 (primary) | 3.0%     | 3.0% | 1         |
</details>

[^1]: *the arithmetic mean of the percent change*
[^2]: *number of relevant changes*

If you disagree with this performance assessment, please file an issue in [rust-lang/rustc-perf](https://github.com/rust-lang/rustc-perf/issues/new).



@rustbot label: -perf-regression
